pkg/agent/flexvolume/controller.go

```diff
@@ -108,44 +108,10 @@ func (c *FlexvolumeController) Attach(attachOpts AttachOptions, devicePath *stri
 				return fmt.Errorf("failed to create volume CRD %s. %+v", crdName, err)
 			}
 			// Some other attacher beat us in this race. Kubernetes will retry again.
-			return fmt.Errorf("failed to attach volume %s. Volume is already attached by a different pod", crdName)
+			return fmt.Errorf("failed to attach volume %s for pod %s/%s. Volume is already attached by a different pod", crdName, attachOpts.PodNamespace, attachOpts.Pod)
 		}
 	} else {
 		// Volume has already been attached.
-		// Check if there is already an attachment with RW.
-		index := getPodRWAttachmentObject(volumeattachObj)
```

Comment:
help me understand the root cause a little better (it would also be useful for you to write up the root cause summary in the issue also: #1082)

- at attachment was attempting for a given pod on a given node
- we created an attachment CRD to capture that pod and node
- the `rbd map` hung, so we cancelled it with SIGINT and failed the attach/mount request
- kubernetes asked our driver to retry the operation.  we check the CRD and see that it's already attached for a pod that exists.
- however, the pod is actually the very same pod that we are trying to attach for on this retry, so we are failing incorrectly.

so the fix here is to instead, once we see that the CRD says this PVC is already attached somewhere, to search the attachments by mount dir.  If we find a matching mount dir, then we proceed on with the rest of the attach flow.  We know this is safe, because according to the CRD, this pod/node/PVC has already been cleared for attaching by our fencing logic.

does that about sum it up?

Reply:
Correcto

Reply:
please add a summary like this to #1082 for posterity